lib/HLSL/HLOperationLower.cpp

```diff
@@ -3586,7 +3586,89 @@ Value *GenerateRawBufLd(Value *handle, Value *bufIdx, Value *offset,
 static Value* TranslateRawBufVecLd(Type* VecEltTy, unsigned VecElemCount,
   IRBuilder<>& Builder, Value* handle, hlsl::OP* OP, Value* status,
   Value* bufIdx, Value* baseOffset, const DataLayout& DL,
-  std::vector<Value*>& bufLds, unsigned baseAlign, bool isScalarTy = false);
+  std::vector<Value*> &bufLds, unsigned actualAlign, unsigned baseAlign, bool isScalarTy = false);
+
+
+static DxilResourceProperties GetResourcePropertyFromHandleCall(const hlsl::HLModule *M, CallInst *handleCall) {
+
+  DxilResourceProperties RP;
+
+  Function *F = handleCall->getCalledFunction();
+  hlsl::HLOpcodeGroup group = hlsl::GetHLOpcodeGroup(F);
+  if (group == hlsl::HLOpcodeGroup::HLAnnotateHandle) {
+    if (Constant *constProps = dyn_cast<Constant>(handleCall->getArgOperand(2))) {
+      RP = resource_helper::loadPropsFromConstant(*constProps);
+    }
+  }
+  return RP;
+}
+
+static Constant* GetRawBufLdStAlignmentInfo(IRBuilder<> &Builder, Value *handle, Type *EltTy,
+  Value *index, Value *offset, Constant *alignment) {
+
+  // If we don't have a constant offset for the element, then give up and
+  // return the default alignment -- i.e. element's allocation size in bytes.
+  if (!offset || !isa<ConstantInt>(offset)) {
```

Comment:
If this can't provide better alignment when offset is not constant, it's not going to be that useful.  The back-end can already easily calculate the alignment of a constant.